include/fuzzuf/optimizer/pso.hpp

```diff
@@ -0,0 +1,229 @@
+#define FUZZUF_INCLUDE_OPTIMIZER_PSO_HPP
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <functional>
+#include <random>
+
+#include "fuzzuf/optimizer/optimizer.hpp"
+#include "fuzzuf/optimizer/pso.hpp"
+#include "fuzzuf/optimizer/store.hpp"
+#include "fuzzuf/utils/common.hpp"
+#include "fuzzuf/utils/random.hpp"
+
+// for Particle Swarm Optimization
+
+namespace fuzzuf::optimizer {
+
+namespace keys {}
+
+template <size_t Dimension, size_t ParticleNum>
+class PSO;
+
+template <size_t Dimension>
+class Particle {
+ public:
+  Particle();
+  ~Particle();
+
+  template <size_t _Dimension, size_t ParticleNum>
+  friend class PSO;
+
+ protected:
+  std::array<double, Dimension> position;
+  double fitness;
+  std::array<double, Dimension> velocity;
+  std::array<double, Dimension> best_position;
+  double best_fitness;
+};
+
+template <size_t Dimension, size_t ParticleNum>
+class PSO : public Optimizer<std::array<double, Dimension>> {
+ public:
+  PSO(double min_position, double max_position, double min_velocity,
+      double max_velocity, double w = 0.729, double c1 = 1.49445,
+      double c2 = 1.49445);
+  ~PSO();
+
+  void Init();
+  std::array<double, Dimension> GetCurParticle();
+  void SetScore(double);
+  std::array<double, Dimension> CalcValue() override;  // return global best
+  void UpdateLocalBest();
+  void UpdateGlobalBest();
+
+ protected:
+  void UpdatePositions();
+  void UpdateVelocities();
+
+  size_t idx = 0;
+  std::uint64_t time = 0;
+  std::array<Particle<Dimension>, ParticleNum> swarm;
+
+  // global best
+  std::array<double, Dimension> best_position;
+  double best_fitness;
+
+  // constraints
+  double min_position;
+  double max_position;
+  double min_velocity;
+  double max_velocity;
+
+  // parameters
+  double w;   // inertia weight
+  double c1;  // cognitive weight
+  double c2;  // social weight
+
+  bool opt_minimize = true;
+};
+
+template <size_t Dimension>
+Particle<Dimension>::Particle() {
+  position.fill(0);
+  velocity.fill(0);
+  best_position.fill(0);
+  best_fitness = 0;
+};
+
+template <size_t Dimension>
+Particle<Dimension>::~Particle(){};
+
+template <size_t Dimension, size_t ParticleNum>
+PSO<Dimension, ParticleNum>::PSO(double min_position, double max_position,
+                                 double min_velocity, double max_velocity,
+                                 double w, double c1, double c2)
+    : min_position(min_position),
+      max_position(max_position),
+      min_velocity(min_velocity),
+      max_velocity(max_velocity),
+      w(w),
+      c1(c1),
+      c2(c2) {}
+
+template <size_t Dimension, size_t ParticleNum>
+PSO<Dimension, ParticleNum>::~PSO() {}
+
+template <size_t Dimension, size_t ParticleNum>
+void PSO<Dimension, ParticleNum>::Init() {
+  for (auto& p : swarm) {
+    for (auto& pos : p.position)
+      pos = fuzzuf::utils::random::Random<double>(min_position, max_position);
+    p.velocity.fill(0);
+  }
+
+  idx = 0;
+  time = 0;
+}
```

Comment:
Why not just do this in the constructor?

Reply:
fixed in b6207e50d21b625fb526c646059baf36fe5a43c4